packages/eventual-send/src/index.js

```diff
@@ -71,33 +68,109 @@ export function makeHandledPromise() {
+      // We propagate p and remove all other pending handlers
       // upstream.
       // Note that everything except presences is covered here.
       while (target !== p) {
         const parent = forwardedPromiseToPromise.get(target);
         forwardedPromiseToPromise.set(target, p);
-        promiseToUnsettledHandler.delete(target);
+        promiseToPendingHandler.delete(target);
         target = parent;
       }
     }
     return target;
   }
 
-  // This special handler accepts Promises, and forwards
-  // handled Promises to their corresponding fulfilledHandler.
+  /**
+   * This special handler accepts Promises, and forwards
+   * handled Promises to their corresponding fulfilledHandler.
+   *
+   * @type {Required<EHandler<any>>}
+   */
   let forwardingHandler;
   let handle;
 
-  function HandledPromise(executor, unsettledHandler = undefined) {
-    if (new.target === undefined) {
-      throw new Error('must be invoked with "new"');
+  /**
+   * @param {string} handlerName
+   * @param {EHandler<any>} handler
+   * @param {string} operation
+   * @param {any} o
+   * @param {any[]} opArgs
+   * @param {Promise<unknown>} [returnedP]
+   * @returns {any}
+   */
+  const dispatchToHandler = (
+    handlerName,
+    handler,
+    operation,
+    o,
+    opArgs,
+    returnedP,
+  ) => {
+    let actualOp = operation;
+    const triedOperations = [operation];
+    const isSendOnly = actualOp.endsWith('SendOnly');
+    const makeResult = result => (isSendOnly ? undefined : result);
+
+    if (isSendOnly) {
+      // We don't specify the resulting promise if it is sendonly.
+      returnedP = undefined;
+    }
+
+    if (isSendOnly && typeof handler[actualOp] !== 'function') {
+      // Substitute for sendonly with the corresponding non-sendonly operation.
+      actualOp = actualOp.slice(0, -'SendOnly'.length);
+      triedOperations.push(actualOp);
+    }
+
+    // Fast path: just call the actual operation.
+    if (typeof handler[actualOp] === 'function') {
+      return makeResult(handler[actualOp](o, ...opArgs, returnedP));
     }
+
+    if (actualOp === 'applyMethod') {
+      // Compose a missing applyMethod by get followed by applyFunction.
+      const [prop, args] = opArgs;
+      const getResultP = handle(o, 'get', [prop], undefined);
+      return makeResult(handle(getResultP, 'applyFunction', [args], returnedP));
+    }
+
+    if (actualOp === 'applyFunction') {
+      actualOp = 'applyMethod';
+      triedOperations.push(actualOp);
+      if (typeof handler[actualOp] === 'function') {
+        // Downlevel a missing applyFunction to applyMethod with undefined name.
+        const [args] = opArgs;
+        return makeResult(handler[actualOp](o, undefined, [args], returnedP));
+      }
+    }
+
+    const { details: X, quote: q } = assert;
+    assert.fail(
+      X`${q(handlerName)} is defined but has no methods needed for ${q(
+        operation,
+      )} (has ${q(Object.keys(handler).sort())})`,
```

Comment:
Given that this one is only about handler names, I suppose its intolerance of symbols is ok. This `q` is tolerant of symbols, but `sort` is not. Which isn't yet an issue anyway because you're using `Object.keys` which omits symbols, which seems fine here, but it also omits non-enumerable properties.

```suggestion
      )} (has ${q(Reflect.ownKeys(handler).map(String).sort())})`,
```

where the optional `.map(String)` would make it tolerant of symbols, which seems like nice future proofing.

However, both here and the other places where such diagnostics appear, we have another problem. This is external proposed API for normal JS users. Some of their objects will inherit their methods, whether because they're using classes, or just manually using prototype inheritance. I don't have any decent suggestions for how to generalize such diagnostics to cover those cases. I'm ok not doing so if we leave behind an explanatory TODO comment and/or file an issue.

Reply:
My guess is that `triedOperations` is supposed to make this diagnostic more informative.